src/main/java/uk/gov/ons/ctp/response/collection/exercise/validation/ValidateSampleUnits.java

```diff
@@ -84,25 +84,36 @@ public void validateSampleUnits() {
     // the sample service are state transitioned once they've got all their sample units.
     updateExecutingCollectionExercises();
 
-    Map<CollectionExercise, Optional<UUID>> collectionInstrumentMap = new HashMap<>();
+    // This is a hand-rolled cache which stores all the collection instruments which are needed
+    // to be added onto the samples which are being 'validated'. The map allows a collection
+    // exercise and a sample form type to be mapped to the correct collection instrument. It
+    // provides a very significant performance boost when 'validating' large samples (e.g. 100k+)
+    Map<CollectionExercise, Map<String, Optional<UUID>>> collectionInstrumentMap = new HashMap<>();
```

Comment:
A Map of Maps feels like it's going to be a huge memory hog with large samples - have you tested it out?

Reply:
Usually there will only be one sample file being processed at any one time with a handful of form types. Because the sample is generally only for one collex, about the most this map of maps would ever hold would be about 30 UUIDs (e.g. MBS)